### ISSUE TITLE
Using a notebook instead of a folder to store the templates can cause various issues

### ISSUE THREAD

**Issue by laurent22** [joplin/plugin-templates#12]
I've been thinking about how we'd now be using a notebook to store the templates, instead of a folder on the hard drive. I can think of many possible issues with this:

- People will delete the template folder. How does your code handle this?

- When setting up a new client, your plugin asks to select a template folder but it hasn't been synced yet.

- Since it's not present, they might ask your plugin to create it and end up with multiple template notebooks each used by a different device.

- People will export the template folder as jex, then import it in a new client, and wonder why your plugin is not using that folder

- They will manually create a folder named "Templates" and wonder why your plugin doesn't pick it up.

- Users will share the Templates notebook using Joplin Server, but the recipient won't be able to use the templates.

I think there are ways to improve this but overall I'm not sure using a template notebook is a good solution. The main problem is that we have no concept of special setting notebooks (there's a thread about this on the forum), which means users can do anything with the notebook and break your plugin behaviour very easily.

So I'm not sure what's your take on this but I get the feeling we've changed a simple solution with no ambiguity (a simple "templates" folder) to one that's quite fragile.

**Comment by nishantwrp** [comment]
> People will delete the template folder. How does your code handle this?

The template selector then will only show the templates that use a template tag. And on restarting the application, the plugin will show the same prompt to select or create a templates notebook.

> When setting up a new client, your plugin asks to select a template folder but it hasn't been synced yet. Since it's not present, they might ask your plugin to create it and end up with multiple template notebooks each used by a different device.

The users have an option to change the templates notebook.

> People will export the template folder as jex, then import it in a new client, and wonder why your plugin is not using that folder

Again they can set that new folder to be the templates notebook.

> They will manually create a folder named "Templates" and wonder why your plugin doesn't pick it up.

Why will they? They'll already be having a templates folder. 

> Users will share the Templates notebook using Joplin Server, but the recipient won't be able to use the templates.

Actually, I'm not familiar with sharing notebooks using Joplin Server. So, no comments on this.

> I think there are ways to improve this but overall I'm not sure using a template notebook is a good solution. The main problem is that we have no concept of special setting notebooks (there's a thread about this on the forum), which means users can do anything with the notebook and break your plugin behaviour very easily. So I'm not sure what's your take on this but I get the feeling we've changed a simple solution with no ambiguity (a simple "templates" folder) to one that's quite fragile.

I don't think it would be ambiguous, once users know that there's one notebook that plugin treats as a templates notebook. They can always change it. And can create templates and other notebooks using a `template` tag.

Also, personally, It felt a little weird to me too, that templates are stored in a directory on my hard drive and not synced. And I don't know what format do I save templates, what's the `Refresh templates` button for etc. And most uses as we know as not programmers so, transferring templates from one machine to another would not be easy for them.

Also, as a poll conducted [here](https://discourse.joplinapp.org/t/opinion-on-the-templates-should-be-organized/17660/2) only 15% of users preferred the existing way. 

What do you think?

**Comment by laurent22** [comment]
I don't think with 26 users the poll is significant, not to mention that it's biased since you mainly list the possible advantages of having a template notebook but none of the drawbacks.

As I said, I know there are ways around these issues but we need to make it user friendly:

- Restarting the app is not good, especially when the user doesn't know they have to do it.

- There will be no indication in the sidebar about which notebook is currently being used for the templates. They might be one or multiple "Templates" notebook but it won't be clear which one is active.

- "Why will they?" is hopeful thinking - if they can **they will** and even if it's only 1% of users it's still high enough if the plugin is widely used.

>  I don't know what format do I save templates, what's the Refresh templates button for etc

That's part of the things you can improve.

I don't like the current implementation overall. It's a crucial part of the plugin and it doesn't look like you put much thought into it.

**Comment by nishantwrp** [comment]
> Restarting the app is not good, especially when the user doesn't know they have to do it.

It's not necessary for them to restart the app. They can select the option to change templates notebook and they again have an option to do create or choose the templates notebook.

> There will be no indication in the sidebar about which notebook is currently being used for the templates. They might be one or multiple "Templates" notebook but it won't be clear which one is active.

From the plugin end, I best we can do is create an option to know which templates notebook is being used as templates notebook.

> "Why will they?" is hopeful thinking - if they can they will and even if it's only 1% of users it's still high enough if the plugin is widely used.

Let's say they do, they know that they've selected a templates notebook initially and they have an option to change it. 

> It's a crucial part of the plugin and it doesn't look like you put much thought into it.

Actually, this was the part I put the most effort and thought into while thinking of the UX, Ease of Use, Migration, taking input from the community, etc. It feels bad to know that you think this was not at all planned. 

> I don't like the current implementation overall.

Anyways, If you think we need to shift to the earlier implementation, I can do that. But maybe, before that as I told after #11, there are two more planned PRs (Loading the earlier templates, documentation). And these can be done this week itself. So, maybe the next week I can create a `.jpl` file and ask the community to give feedback on what they think. Thoughts?

**Comment by nishantwrp** [comment]
Also, if interaction with the notebook seems the only and major problem. I have an solution. Let's not use a notebook at all. We can just use a tag called `template`. As it's always unique there will be no issues with this. The earlier templates can be loaded into a new notebook created and a tag called `template` will automatically be applied. The users can then arrange the loaded templates according to their convenience. Thoguhts?

**Comment by CalebJohn** [comment]
I agree with you @nishantwrp about using tags. The only issue with the tagging system I can see is that it breaks the implicit contract with the user that tags are only used for sorting, and they (the user) have 100% control of those tags. That said, it seems justifiable to me because this is plugin, in fact I'm pretty sure there are other plugins that use tags for functionality.

With that in mind I think that you (@nishantwrp) should make a post on the forum updating the plugin spec. But let's wait for Laurent's response first.

>    When setting up a new client, your plugin asks to select a template folder but it hasn't been synced yet.
>    Since it's not present, they might ask your plugin to create it and end up with multiple template notebooks each used by a different device.

Not an issue since all templates will have the tag. Duplicates will have to be remediated manually (won't be an issue in the future). A de-duplication/merging command can easily be added for convenience.

>    People will export the template folder as jex, then import it in a new client, and wonder why your plugin is not using that folder

Works with tags.
>    They will manually create a folder named "Templates" and wonder why your plugin doesn't pick it up.

Not needed anymore.
>    Users will share the Templates notebook using Joplin Server, but the recipient won't be able to use the templates.

Works since template tag will work by name, not id.
> There will be no indication in the sidebar about which notebook is currently being used for the templates. They might be one or multiple "Templates" notebook but it won't be clear which one is active.

Not needed with tags.

**Comment by laurent22** [comment]
I didn't follow this tag feature, is that recent? Indeed a spec that explains how it work would be useful. Also is there something automated about this tag, or is that the user manually adding it or removing it from notes?

Also tag titles aren't unique - there could be multiple "template" tags if they are created separately on multiple devices and then synced.

**Comment by CalebJohn** [comment]
> is there something automated about this tag, or is that the user manually adding it or removing it from notes?

Just when first importing the templates from the old system, otherwise not. 

> Also tag titles aren't unique - there could be multiple "template" tags if they are created separately on multiple devices and then synced.

The current implementation of tag handling (#11) assumes there is exactly one tag with the name "template", but it could easily be extended to use all notes with the tag. Using the name rather than ID is to our benefit because it does away with all the issues of duplicates, and makes the templating What You See is What You Get (e.g. if a note has a "template" tag, you can be sure that it's accessible as a template). Merging duplicated tags would need to be done manually.

**Comment by nishantwrp** [comment]
> Merging duplicated tags would need to be done manually.

Can you explain a bit about merging? 

> Also tag titles aren't unique - there could be multiple "template" tags if they are created separately on multiple devices and then synced.

If this is the case, we can fetch the notes for all the tags named `template`.

**Comment by CalebJohn** [comment]
> Can you explain a bit about merging?

Suppose this plugin is updated to only use tags and not a folder. Then, if this plugin is installed before syncing on multiple computers, the user will end up with duplicates of the "template" tag.  That is, there will be multiple version of the tag with the same name but different IDs. The user will have to manually "merge" the contents of those tags.

**Comment by nishantwrp** [comment]
> Suppose this plugin is updated to only use tags and not a folder. Then, if this plugin is installed before syncing on multiple computers, the user will end up with duplicates of the "template" tag. That is, there will be multiple version of the tag with the same name but different IDs. The user will have to manually "merge" the contents of those tags.

But this will be rare, right? The automation happens only when there are old templates of the native feature.

**Comment by CalebJohn** [comment]
> But this will be rare, right? The automation happens only when there are old templates of the native feature.

Depends on how you define rare. You could make it rare by only triggering your plugin after the first sync, and then forcing a sync after adding templates. Otherwise, as things are now, this is likely to be experienced by the majority of template users. And it will be their first experience with your plugin, so it's worth thinking about.

**Comment by nishantwrp** [comment]
> > But this will be rare, right? The automation happens only when there are old templates of the native feature.
> 
> Depends on how you define rare. You could make it rare by only triggering your plugin after the first sync, and then forcing a sync after adding templates. Otherwise, as things are now, this is likely to be experienced by the majority of template users. And it will be their first experience with your plugin, so it's worth thinking about.

Oh, okay got it. So, can we consider adding an option in the menu for loading old templates instead of doing it automatically? And in the welcome message, we can add a line saying that you can load your previous templates by an option in the menu.

**Comment by CalebJohn** [comment]
> So, can we consider adding an option in the menu for loading old templates instead of doing it automatically?

Using the [onSyncStart](https://joplinapp.org/api/references/plugin_api/classes/joplinworkspace.html#onsynccomplete) event would be better.

> Also, by this, you mean the template users that have templates in multiple Joplin Desktop Clients, right?

Yes.

**Comment by nishantwrp** [comment]
Nice idea, but shouldn't we do it on onSyncFinish?

**Comment by CalebJohn** [comment]
> Nice idea, but shouldn't we do it on onSyncFinish?

Just a typo, you'll see a linked to `onSyncComplete`.

**Comment by laurent22** [comment]
Is there any plan to create a spec for this then? I guess it should say:
- What is the problem
- How is it being solved
- What will be the UI and UX
- What are the known issues with this new implementation

**Comment by nishantwrp** [comment]
It would be great if others can post their opinions too on this feature.

cc: @laurent22 @CalebJohn
